cudax/include/cuda/experimental/__event/event.cuh

```diff
@@ -0,0 +1,162 @@
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _CUDAX_EVENT_DETAIL_H
+#define _CUDAX_EVENT_DETAIL_H
+
+#include <cuda_runtime_api.h>
+// cuda_runtime_api needs to come first
+
+#include <cuda/std/detail/__config>
+
+#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
+#  pragma GCC system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+#include <cuda/std/__cuda/api_wrapper.h>
+#include <cuda/std/__exception/cuda_error.h>
+#include <cuda/std/cstddef>
+#include <cuda/std/utility>
+#include <cuda/stream_ref>
+
+// CUDAX headers here
+#include <cuda/experimental/__detail/utility.cuh>
+#include <cuda/experimental/__event/event_ref.cuh>
+
+#include <cassert>
+
+namespace cuda::experimental
+{
+class timed_event;
+
+/**
+ * @brief An owning wrapper for an untimed `cudaEvent_t`.
+ */
+class event : public event_ref
+{
+  friend class timed_event;
+
+public:
+  /**
+   * @brief Flags to use when creating the event.
+   */
+  enum class flags : unsigned int
+  {
+    none          = cudaEventDefault,
+    blocking_sync = cudaEventBlockingSync,
+    interprocess  = cudaEventInterprocess
+  };
+
+  /**
+   * @brief Construct a new `event` object with timing disabled.
+   *
+   * @throws cuda_error if the event creation fails.
+   */
+  event()
+      : event(static_cast<unsigned int>(cudaEventDisableTiming))
+  {}
+
+  /**
+   * @brief Construct a new `event` object with the specified flags.
+   *
+   * @throws cuda_error if the event creation fails.
+   */
+  explicit event(flags __flags)
+      : event(static_cast<unsigned int>(__flags) | static_cast<unsigned int>(cudaEventDisableTiming))
+  {}
+
+  /**
+   * @brief Construct a new `event` object into the moved-from state.
+   *
+   * @post `get()` returns `cudaEvent_t()`.
+   */
+  explicit constexpr event(uninit_t) noexcept {}
+
+  /**
+   * @brief Move-construct a new `event` object
+   *
+   * @param __other
+   *
+   * @post `__other` is in a moved-from state.
+   */
+  constexpr event(event&& __other) noexcept
+      : event_ref(_CUDA_VSTD::exchange(__other.__event_, {}))
+  {}
+
+  /**
+   * @brief Destroy the `event` object
+   *
+   * @note If the event fails to be destroyed, the error is silently ignored.
+   */
+  ~event()
+  {
+    if (__event_ != nullptr)
+    {
+      [[maybe_unused]] auto __status = ::cudaEventDestroy(__event_);
+    }
+  }
+
+  /**
+   * @brief Construct an `event` object from a native `cudaEvent_t` handle.
+   *
+   * @param __evnt The native handle
+   *
+   * @return event The constructed `event` object
+   *
+   * @note The constructed `event` object takes ownership of the native handle.
+   */
+  _CCCL_NODISCARD static event from_native_handle(::cudaEvent_t __evnt) noexcept
+  {
+    return event(__evnt);
+  }
```

Comment:
Could we just use `__event` That single letter doesnt save much and makes it easier to read
```suggestion
  _CCCL_NODISCARD static event from_native_handle(::cudaEvent_t __event) noexcept
  {
    return event(__event);
  }
```

Reply:
In fact we cannot. I named it that originally but it broke the msvc build. Apparently `__event` is a Word Of Power in Windows-land.

Reply:
Thanks I hate it

Reply:
Could it be named __handle instead, or is it reserved as well?